urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/client/user/User.java

```diff
@@ -0,0 +1,206 @@
+ *
+ * @author Friedemann
+ */
+public class User extends Thread {
+
+	/** The user id given by the client. */
+	private final UUID userID;
+
+	/** The user's browser instance. */
+	private final Browser b;
+
+	/** True if the user acts with default behaviour. */
+	private boolean isActive = true;
+
+	/** The article list. */
+	private ArrayList<String> articleList;
+
+	/** The timestamp the user started. */
+	private Timestamp timeStarted = null;
+
+	/** The time to live in seconds. */
+	private final int timeToLive;
+
+	/** Timer to measure request durations. */
+	private final Timer tmr;
+
+	/** The random. */
+	private final Randomizer random;
+
+	/** The client reference. */
+	private final Client client;
+
+	/**
+	 * Instantiates a new user.
+	 *
+	 * @param userID the user id
+	 */
+	public User(final Client c, final BrowserType type, final int ttl) {
+
+		client = c;
+		timeToLive = ttl;
+
+		// UUID is not thread safe
+		synchronized (client.lf) {
+			userID = UUID.randomUUID();
+		}
+		this.setName("user-" + userID.toString());
+
+		b = new Browser(type, client.testTicket.getWebServerURI(), client.testTicket
+				.getWebSocketURI());
+		tmr = new Timer();
+		random = new Randomizer();
+	}
+
+	/**
+	 * Gets the user id.
+	 *
+	 * @return the user id
+	 */
+	public UUID getUserID() {
+		return userID;
+	}
+
+	/**
+	 * Sets the article list.
+	 *
+	 * @param list the new article list
+	 */
+	public void setArticleList(final ArrayList<String> list) {
+		articleList = list;
+	}
+
+	/**
+	 * Sets the user (in-)active.
+	 *
+	 * @param act
+	 */
+	public void setActive(final boolean act) {
+		isActive = act;
+	}
+
+	/**
+	 * Gets the time the user is running in seconds.
+	 *
+	 * @return the time running
+	 */
+	public int getTimeRunning() {
+
+		if (timeStarted != null)
+			return (int) ((System.currentTimeMillis() - timeStarted.getTime()) / 1000);
+		else
+			return 0;
+
+	}
+
+	/**
+	 * Gets user's time to live in seconds (tab open time)
+	 *
+	 * @return int the time in seconds
+	 */
+	public int getTimeToLive() {
+		return timeToLive;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.lang.Runnable#run()
+	 */
+	@Override
+	public void run() {
+
+		// keep start time
+		if (timeStarted == null)
+		{
+			timeStarted = new Timestamp(System.currentTimeMillis());
+		}
+
+		String nextArticleID;
+
+		while (isActive)
+		{
+			// #1 BLOCKING until all users are initialized
+			/*
+			while((client.isAllUsersInitialized() == false) && isActive)
+			{
+				try {
+					Thread.sleep(500);
+				} catch (final InterruptedException e) {
+					client.cgui.logMessage("Interrupted while waiting for all users! "+e);
+				}
+			}
+			*/
+
+			// #2 Goto sleep while reading (assuming the user is reading the main page first :-)
+			if (isActive)
+			{
+				try
+				{
+					Thread.sleep(random.getInt(client.testTicket.getMinReadingTime(),
+							client.testTicket.getMaxReadingTime()) * 1000);
+
+				} catch (final InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)